### ISSUE TITLE
Errors occurring when using `mock` with selenoid (remote)

### ISSUE THREAD

**Issue by DudaGod** [webdriverio/webdriverio#7086]
**Environment (please complete the following information):**
 - **WebdriverIO version:** 7.7.4
 - **Mode:** Standalone mode
 - **Node.js version:** 12.18.1
 - **NPM version:** 6.14.5
 - **Browser name and version:** Chrome 72
 - **Platform name and version:** Linux

**Describe the bug**
I have selenoid with chrome@72 in some remote machine. I want to mock some requests using method `mock`. 
I write test in which I mock all requests and log all requested urls:
```
const {remote, attach} = require('webdriverio');
let client, client2;

const options = {
    protocol: 'http',
    hostname: 'MY_GRID_HOSTNAME',
    port: 4444,
    path: '/wd/hub',
    automationProtocol: 'webdriver',
    capabilities: {
        browserName: 'chrome',
        browserVersion: '72',
    }
};

(async function () {
    client = await remote(options);
    client2 = await attach(client);

    const mock = await client2.mock('**');

    await client2.url('https://yandex.ru');
    console.log('visited urls:', mock.calls.map(({url}) => url));

    await client2.deleteSession();
})()
    .catch(async (e) => {
        console.log(e);
        await client.deleteSession();
    })
```

First error that I get - `Cannot read property 'selenoid:options' of undefined` from here - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/getPuppeteer.ts#L71.

It means that `client.requestedCapabilities` didn't apply to attached client. And even if I pass it explicitly:
```
client2 = await attach({...client, requestedCapabilities: client.requestedCapabilities});
```

It fix the previous error but a new error will appear:
`Failed to fetch browser webSocket URL from http://localhost:46717/json/version: request to http://localhost:46717/json/version failed, reason: connect ECONNREFUSED 127.0.0.1:46717`

It means that wdio try to connect to local chrome inside `getPuppeteer` method. But I don't understand why should I necessarily pass `selenoid:options` to connect to my remote browser? It looks like that it is necessary for the wdio to understand that a selenoid is being used.

Okay, I fix my capabilities and pass `'selenoid:options: {}` explicitly. And now I get new error - `Unexpected server response: 404`. In full log of error I see that wdio was try to connect to chrome using this url - `ws://localhost:4444/devtools/a92ca733b9d02942f620c51c5fd0570e7ab4f39fdba0b87103f62b468167b25c`. It means that `hostname` from options here - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/getPuppeteer.ts#L73 has incorrect value.

Okay, now I pass my all `options` with `hostname`, `protocol` and etc to attach method explicitly:
```
client2 = await attach({...client, requestedCapabilities: client.requestedCapabilities, ...options});
```

And now I get the last error: `'Fetch.enable' wasn't found`. I found in this issue - https://github.com/puppeteer/puppeteer/issues/4542 that for each chrome should use different version of puppeteer. But when I installed puppeteer-core@chrome-72 in node_modules of webdriverio it didn't fix the problem. Because call of `Fetch.enable` is used inside wdio - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/mock.ts#L159 but it doesn't support in chrome@72.

I write script using puppeter-core@chrome-72 and intercept requests works fine:
```
const {remote, attach} = require('webdriverio');
const puppeteer = require('puppeteer-core');

(async () => {
    const browser = await remote({
        protocol: 'http',
        hostname: 'MY_GRID_HOSTNAME',
        path: '/wd/hub',
        port: 4444,

        logLevel: 'trace',
        automationProtocol: 'webdriver',

        capabilities: {
            browserName: 'chrome',
            browserVersion: '72'
        }
    });

    const devtools = await puppeteer.connect(
        {browserWSEndpoint: `ws://MY_GRID_HOSTNAME:4444/devtools/${browser.sessionId}`}
    );

    let pages = await devtools.pages();

    await pages[0].setRequestInterception(true);
    pages[0].on('request', async (request) => {
        console.info('req.URL:', request.url());
        return request.continue(); // Allow request to continue
    })

    await browser.url('https://yandex.ru');

    await devtools.close();
    await browser.deleteSession();
})().catch((e) => console.error(e));
```

To solve the problem with wdio I began to use chrome@91 and finally everything worked as I expected. Final view of the working script:

```
const {remote, attach} = require('webdriverio');
let client, client2;

const options = {
    protocol: 'http',
    hostname: 'MY_GRID_HOSTNAME',
    port: 4444,
    path: '/wd/hub',
    automationProtocol: 'webdriver',

    capabilities: {
        browserName: 'chrome',
        browserVersion: '91',
        'selenoid:options': {}
    }
};

(async function () {
    client = await remote(options);
    client2 = await attach({...client, requestedCapabilities: client.requestedCapabilities, ...options});

    const mock = await client2.mock('**');

    await client2.url('https://yandex.ru');
    console.log('visited urls:', mock.calls.map(({url}) => url));

    await client2.deleteSession();
})()
    .catch(async (e) => {
        console.log(e);
        await client.deleteSession();
    })
```

**Comment by BorisOsipov** [comment]
>But I don't understand why should I necessarily pass selenoid:options to connect to my remote browser? 

@DudaGod reasonable question. But answer the question - how can we detect that the session is selenoid session? Selenoid proxies all requests\responses from\to chromedriver binary - so we don't know how detect Selenoid session without requested vendor capabilities.

**Comment by BorisOsipov** [comment]
@DudaGod FMPOV there are several question\issues in your issue. Could you please recap what you expect for each point? 
I think we should disscuss each point here
 and split this issue to few small feature requests then.
 
Also, thank you so much for detailed bug reports!

**Comment by DudaGod** [comment]
> @DudaGod reasonable question. But please can you answer the question - how can we detect that the session is selenoid session? Selenoid proxies all requests\responses from\to chromedriver binary - so we don't know how detect Selenoid session without requested vendor capabilities.

To my sorrow, I don't see any other options at the moment. Maybe we should try to speak with guys from selenoid.

> Could you please recap what you expect for each point?

Yeah, of course.

Lets go:
- > First error that I get - Cannot read property 'selenoid:options' of undefined from here - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/getPuppeteer.ts#L71.

Here I would expect that `client.requestedCapabilities` will be used under the hood when I pass browser instance to `attach` method.

- > It fix the previous error but a new error will appear:
Failed to fetch browser webSocket URL from http://localhost:46717/json/version: request to http://localhost:46717/json/version failed, reason: connect ECONNREFUSED 127.0.0.1:46717

Here i would expect to see some hints inside error which helps to understand where is the problem. For example if you use selenoid add `selenoid:options: {}` to `capabilities`. It would seem convenient for every error to have a doc link where we can describe the different options in more detail.

- > And now I get new error - Unexpected server response: 404. In full log of error I see that wdio was try to connect to chrome using this url - ws://localhost:4444/devtools/a92ca733b9d02942f620c51c5fd0570e7ab4f39fdba0b87103f62b468167b25c. It means that hostname from options here - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/getPuppeteer.ts#L73 has incorrect value.

Here I would expect that `client.options` will be used under the hood as well as `client.requestedCapabilities` in the first error.

- > And now I get the last error: 'Fetch.enable' wasn't found. I found in this issue - puppeteer/puppeteer#4542 that for each chrome should use different version of puppeteer. But when I installed puppeteer-core@chrome-72 in node_modules of webdriverio it didn't fix the problem. Because call of Fetch.enable is used inside wdio - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/mock.ts#L159 but it doesn't support in chrome@72.

Here I would expect that usage restrictions of CDP in remote browsers to be described in wdio documentation. Maybe that error should also be accompanied by a link to the documentation.

**Comment by christian-bromann** [comment]
> Here I would expect that `client.requestedCapabilities` will be used under the hood when I pass browser instance to `attach` method.

Can you provide a minimal reproducible example for this? It seems that we don't correctly create an attach instance.

> Here i would expect to see some hints inside error which helps to understand where is the problem.

I am afraid this is connected to the first problem.

> Here I would expect that usage restrictions of CDP in remote browsers to be described in wdio documentation.

That is impossible. You are using a quite old version of Chrome that doesn't not have this domain implemented. I don't see a way to document such things as we can't document all changes in individual browser separately.

**Comment by DudaGod** [comment]
> Can you provide a minimal reproducible example for this? It seems that we don't correctly create an attach instance.

Yes, it looks like. My minimal reproducible example was added to issue description, but I can duplicate it:
```
const {remote, attach} = require('webdriverio');
let client, client2;

const options = {
    protocol: 'http',
    hostname: 'MY_GRID_HOSTNAME',
    port: 4444,
    path: '/wd/hub',
    automationProtocol: 'webdriver',
    capabilities: {
        browserName: 'chrome',
        browserVersion: '91',
        'selenoid:options': {}
    }
};

(async function () {
    client = await remote(options);
    client2 = await attach(client);

    const mock = await client2.mock('**');

    await client2.url('https://yandex.ru');
    console.log('visited urls:', mock.calls.map(({url}) => url));

    await client2.deleteSession();
})()
    .catch(async (e) => {
        console.log(e);
        await client.deleteSession();
    });
```

**Comment by christian-bromann** [comment]
@DudaGod this is a bug as `requestedCapabilities` is not properly propagated to the `attachToSession` method. Seems that deconstructing [`...attachOptions`](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/index.ts#L84) doesn't work as expected. Any contributions to this would be appreciated.

**Comment by christian-bromann** [comment]
Fixed in #7155
